docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,410 @@
+
+While DHCP is a viable solution for dynamic IP assignment which also enables scaling, it can cause problems in conjunction with CAPI. Especially in smaller networks rolling cluster upgrades can exhaust the network in use. When multiple machines are replaced in quick succession, each of them will get a new DHCP lease. Unless the lease time is very short, at least twice as many IPs as the maximum number of nodes has to be allocated to a cluster.
+
+Metal3 has an ip-address-manager component that allows for in-cluster management of IP addresses through custom resources, allowing to avoid DHCP. CAPV allows to ommit the address from the template while having DHCP disabled, and will wait until it is set manually, allowing to implement custom controllers to take care of IP assignments. At DTAG we've extended metal3's ip-address-manager and wrote a custom controller for CAPV to integrate both with Infoblox, our IPAM solution of choice. The CAPV project has shown interest, and there has been a proposal to allow integrating metal3's ip-address-manager with external systems.
+
+All on-premise providers have a need for IP Address Management, since they can't leverage SDN features that are available in clouds such as AWS, Azure or GCP. We therefore propose to add an API contract to CAPI, which allows infrastructure providers to integrate with different IPAM systems. A similar approach to Kubernetes' PersistentVolumes should be used where infrastructure providers create Claims that reference a specific IP Pool, and IPAM providers fulfill claims to Pools managed by them.
+
+### Goals
+
+- Define an API contract that
+    - allows infrastructure providers to dynamically allocate and release addresses
+    - allows to write IPAM providers to integrate with any IPAM solution
+    - supports both IPv4 and IPv6
+
+### Non-Goals/Future Work
+
+- IPAM providers should not be added to CAPI directly, but live as individual projects instead.
+- IPAddress and IPAddressPool will only have limited validation (e.g. that the referenced Pool exists and that IP address strings are valid addresses). Advanced validation (whether an IP is actually part of the correct subnet for example) needs to be handled by IPAM providers.
+
+## Proposal
+
+The chosen approach was heavily inspired by metal3's ip-address-manager, which in turn seems to have been inspired by Kubernetes' PersistentVolumes. The proposal is twofold and consists of the API contract that should be added to CAPI and two examples to better illustrate how the contract is used by IPAM and infrastructure providers.
+
+### IPAM API Contract
+
+The IPAM API contract provides a generic interface between CAPI infrastructure providers and to-be-developed IPAM providers. Any intersted infrastructure is supposed to be able to request IP Addresses from a defined pool of addresses. IPAM providers should then be able to find and fulfil such requests using any custom logic to do so.
+
+The contract consists of three resources:
+
```

Comment:
do we also need to elaborate the contractual conditions between the infraMachine CR and the IPAddressClaim?

Reply:
They are mentioned further down in the consumption section, but they should probably be explicitly listed in the contract.